src/adapter/src/coord/command_handler.rs

```diff
@@ -888,13 +888,21 @@ impl Coordinator {
             // Acquire read holds _before_ we determine the least valid read.
             // Otherwise, we're not guaranteed that the since frontier doesn't
             // advance forward from underneath us.
+            let read_holds = self
+                .acquire_read_holds(timestamp, &ids, false)
+                .expect("precise==false, so acquiring read holds always succeeds");
```

Comment:
We don't have to do it for this PR, but we should probably split `acquire_read_holds` into two methods. One that returns a `Result` and one that doesn't.

Reply:
agreed! I'd thought that they would neatly partition into two groups where we call it either with `timestamp = MIN, precise = false` or `timestamp = <some ts>, precise = true`, but that's not the case. Should still split them though, because this boolean is hard to reason about at callsites.

Reply:
Also, how about fusing `acquire_read_holds` and `least_valid_read` into one function? There are now many call sites where `least_valid_read` immediately follows `acquire_read_holds`.

So, how I imagine it is that the current `least_valid_read` function could be renamed to `least_valid_read_from_read_holds`, and a new function could be named simply `least_valid_read`, which would call `acquire_read_holds` and then `least_valid_read_from_read_holds`. And then most call sites could be just calling `least_valid_read`, except for some tricky call sites that pass in read holds from afar so would call `least_valid_read_from_read_holds`.

Reply:
I once that a version like this (that combined getting read holds and determining a least valid read) in my big refactoring branch but it felt unwieldy. I'd like to keep them separate for now even though most call-sites do actually do both things.

We can always revisit that in the future, IMO.